src/pages/workspaces/workspace/delete-workspace-modal.state.ts

```diff
@@ -0,0 +1,107 @@
+import _ from 'lodash/fp'
+import { useState } from 'react'
+import { Ajax } from 'src/libs/ajax'
+import { reportError, withErrorReportingInModal } from 'src/libs/error'
+import { useCancellation, useOnMount } from 'src/libs/react-utils'
+import { getUser } from 'src/libs/state'
+import * as Utils from 'src/libs/utils'
+import { BaseWorkspace, isGoogleWorkspace, WorkspaceInfo } from 'src/libs/workspace-utils'
+import { isResourceDeletable } from 'src/pages/workspaces/workspace/analysis/runtime-utils'
+
+
+export interface DeleteWorkspaceState {
+  loading: boolean
+  deleting: boolean
+  isDeleteDisabledFromResources: boolean
+  workspaceBucketUsageInBytes: any
+  deletableApps: any
+  nonDeletableApps: any
+  collaboratorEmails: any
+  hasApps: () => boolean
+  deleteWorkspace: () => void
+  controlledResourcesExist: boolean
+}
+
+
+export const useDeleteWorkspaceState = (workspace: BaseWorkspace, onDismiss: () => void, onSuccess: () => void) : DeleteWorkspaceState => {
+  const [deleting, setDeleting] = useState(false)
+  const [loading, setLoading] = useState(false)
+  const [apps, setApps] = useState()
+  const [collaboratorEmails, setCollaboratorEmails] = useState()
+  const [workspaceBucketUsageInBytes, setWorkspaceBucketUsageInBytes] = useState()
+  const [controlledResourcesExist, setControlledResourcesExist] = useState(false)
+  const [deletableApps, nonDeletableApps] = _.partition(isResourceDeletable('app'), apps) as any // TODO fix this
+
+  const workspaceInfo: WorkspaceInfo = workspace.workspace
+  const signal = useCancellation()
+
+  useOnMount(() => {
+    const load = _.flow(
+      withErrorReportingInModal('Error checking workspace resources', onDismiss),
+      Utils.withBusyState(setLoading)
+    )(async () => {
+      if (isGoogleWorkspace(workspace)) {
+        const [currentWorkspaceAppList, { acl }, { usageInBytes }] = await Promise.all([
+          Ajax(signal).Apps.listWithoutProject({ role: 'creator', saturnWorkspaceName: workspaceInfo.name }),
+          Ajax(signal).Workspaces.workspace(workspaceInfo.namespace, workspaceInfo.name).getAcl(),
+          Ajax(signal).Workspaces.workspace(workspaceInfo.namespace, workspaceInfo.name).bucketUsage()
+        ])
+        setApps(currentWorkspaceAppList)
+
+        // @ts-ignore
+        setCollaboratorEmails(_.without([getUser().email], _.keys(acl)))
+        setWorkspaceBucketUsageInBytes(usageInBytes)
+      } else {
+        const currentWorkspaceAppList = await Ajax(signal).Apps.listAppsV2(workspaceInfo.workspaceId)
+        // temporary hack to prevent orphaning resources on Azure:
+        // change each app to status: 'disallow' which will cause this modal to think they are undeletable
+        const hackedAppList = _.map(_.set('status', 'disallow'), currentWorkspaceAppList)
+        // TODO fix this
+        // @ts-ignore
+        setApps(hackedAppList)
```

Comment:
I am uncertain what the best/most-idiomatic approach is here, would love some opinions from more experienced TypeScripters. The issue seems to be that the call above to lodash-fp is giving back an `unknown[]` which is not assignable to the `undefined` state var for `currentWorkspaceAppList`.

Reply:
My recommendation would be to add types to Ajax.Apps.listAppsV2() method so that the code on like 59 knows that it is returning a Promise<DeleteWorkspaceModalLeoApp>.  Hopefully getting currentWorkspaceAppList typed through that effort will then be sufficient to have types flow thru _.map and _.set on like 63 to infer the correct type to hackedAppList.   Then we get the added bonsus of one more Ajax sub-area being converted to at least partially well typed TS.

Looks like Apps subarea is already a seperate .js file in Ajax directory...  Convert to TS and define DeleteWorkspaceModalLeoApp, uh LeoApp type there....worth a chat with IA team to get synced up.

Reply:
I will reach out to the IA team about typescriptifying that code, I am going to proceed with this approach for now to keep this PR from sprawling.